DREAM/Optimization/tsgGradientDescent.hpp

```diff
@@ -0,0 +1,154 @@
+ *
+ * \par class GradientDescentState
+ * An object of this class represents the components of a gradient descent run and all key parts of the swarm can be read through
+ * member methods. Methods are also provided to modify a subset of its parts.
+ *
+ * \par Constructors and Copy/Move assignment
+ * All constructors require information about an initial candidate point and initial stepsize. A method is available to retrieve
+ * the number of dimensions. No default constructor is available for this class, but copy/move constructors and assignment operator=
+ * overloads are available. The number of dimensions \b cannot be modified.
+ * - GradientDescentState()
+ * - getNumDimensions()
+ *
+ * \par Set and Get Candidate and Stepsize
+ * Throughout the run of gradient descent, the candidate point and stepsize may change. Methods are available to retrieve these
+ * data and set them manually.
+ * - getCandidate(), setCandidate()
+ * - getStepsize(), setStepsize()
+ *
+ * \par Gradient Descent Algorithm
+ * The gradient descent optimization algorithm can be used on a gradient descent algorithm with the goal of minimizing a particular
+ * objective function. The current candidate point, with respect to the given objective function, can be obtained from the state
+ * using the getCandidate() method.
+ * - GradientDescent()
+ *
+ * \par
+ * More information about the gradient descent algorithm can be found in the following paper:
+ *
+ * \par
+ * > Nesterov, Y. (2013). Gradient methods for minimizing composite functions. <em>Mathematical programming</em>, 140(1), 125-161.
+ *
+ */
+class GradientDescentState {
+  public:
+    //! \brief The default constructor is allowed.
+    GradientDescentState() = delete;
+    //! \brief Constructor for a gradient descent state with the initial candidate and stepsize.
+    GradientDescentState(const std::vector<double> &x, const double stepsize);
+    //! \brief Copy constructor.
+    GradientDescentState(const GradientDescentState &source) = default;
+    //! \brief Move constructor.
+    GradientDescentState(GradientDescentState &&source) = default;
+
+    //! \brief Move assignment.
+    GradientDescentState& operator=(GradientDescentState &&source) = default;
+
+    //! \brief Return the number of dimensions.
+    inline int getNumDimensions() const {return num_dimensions;}
+    //! \brief Return the stepsize.
+    inline double getStepsize() const {return stepsize;}
+    //! \brief Return the current candidate point.
+    inline void getCandidate(double x[]) const {std::copy_n(candidate.begin(), num_dimensions, x);}
+    inline std::vector<double> getCandidate() const {return candidate;}
+
+    //! \brief Set the stepsize.
+    inline void setStepsize(const double ss) {stepsize = ss;}
+    //! \brief Set the current candidate point.
+    inline void setCandidate(const double x[]) {std::copy_n(x, num_dimensions, candidate.begin());}
+    inline void setCandidate(const std::vector<double> &x) {
```

Comment:
Looking at the state as a whole, we can probably implement the whole thing with:
```
using GradientDescentState = std::vector<double>;
```